rtCommon/bidsArchive.py

```diff
@@ -0,0 +1,825 @@
+
+        Returns:
+            A list of images matching the provided entities (empty if there are
+            no matches, and containing at most a single image if an exact match
+            is requested).
+
+        Examples:
+            >>> archive = BidsArchive('/path/to/archive')
+
+            Using a dictionary to provide target entities.
+
+            >>> entityDict = {'subject': '01', 'datatype': 'func'}
+            >>> images = archive.getImages(**entityDict)
+
+            Using keyword arguments to provide target entities.
+
+            >>> images = archive.getImages(subject='01', datatype='func')
+
+            Accessing properties of the image.
+
+            >>> image = images[0]
+            >>> print(image.get_image()
+            (64, 64, 27, 3)
+            >>> print(image.path)
+            /tmp/archive/func/sub-01_task-test_bold.nii
+            >>> print(image.filename)
+            sub-01_task-test_bold.nii
+
+            An exact match must have exactly the same entities; since images
+            must also have the task entity in their filename, the above
+            entityDict will yield no exact matches in the archive.
+
+            >>> images = archive.getImages(entityDict, matchExact=True)
+            ERROR "No images were an exact match for: {'subject': '01',
+            'datatype': 'func'}"
+            >>> print(len(images))
+            0
+        """
+        # Validate image extension specified
+        extension = entities.pop('extension', None)
+        if extension is not None:
+            if extension != '.nii' and extension != '.nii.gz':
+                raise ValueError('Extension for images must be either .nii or '
+                                 '.nii.gz')
+
+        results = self.data.get(**entities)
+        results = [r for r in results if type(r) is BIDSImageFile]
+
+        if len(results) == 0:
+            logger.debug(f"Found no images with all entities: {entities}")
+            return []
+        elif matchExact:
+            for result in results:
+                # Only BIDSImageFiles are checked, so extension is irrelevant
+                result_entities = result.get_entities()
+                result_entities.pop('extension', None)
+
+                if result_entities == entities:
+                    return [result]
+
+            logger.debug(f"Found no images exactly matching: {entities}")
+            return []
+        else:
+            return results
+
+    def _ensurePathExists(self, relPath: str):
+        """
+        Ensures the provided directory path exists in the dataset, creating
+        directories if needed.
+
+        Args:
+            relPath: Path to ensure existence of, relative to directory root.
+                All parts of the path are assumed to be directories.
+
+        Returns:
+            True if a layout update is needed, false otherwise.
+        """
+        if self.dirExists(relPath):
```

Comment:
GW comment: seems counter-intuitive to return false when it exists. Maybe this function shouldn't return anything?